File: 算法/代码随想录算法PDF/00调试源代码/501.二叉搜索树中的众数.py
#
# @lc app=leetcode.cn id=501 lang=python3
#
# [501] 二叉搜索树中的众数
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        maxCount = 0
        count = 0
        pre = None
        res = []

        st = []
        while root or st:
            if root:
                st.append(root)
                root = root.left
            else:
                root = st.pop()

                if not pre:
                    count = 1
                elif pre.val == root.val:
                    count += 1
                else:
                    count = 1
                pre = root

                if count == maxCount: res.append(root.val)
                if count > maxCount:
                    maxCount = count
                    res = []
                    res.append(root.val)

                root = root.right
        return res
# @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/501.二叉搜索树中的众数.cpp
/*
 * @lc app=leetcode.cn id=501 lang=cpp
 *
 * [501] 二叉搜索树中的众数
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> findMode(TreeNode* root) {
        TreeNode* pre = nullptr;  // 记录前一个节点
        int maxCount = 0;
        int count;
        vector<int> res;
        stack<TreeNode*> st;
        int minDiff = INT_MAX;
        while (root || !st.empty()) {
            if (root) {
                st.push(root);
                root = root->left;
            } else {
                root = st.top(); st.pop();

                if (!pre) {  // 统计当前遍历元素值的频次
                    count = 1;
                } else if (pre->val == root->val) {
                    count++;
                } else {
                    count = 1;
                }
                pre = root;

                if (count == maxCount) res.push_back(root->val);  // 验证是否为最大频次
                if (count > maxCount) {
                    maxCount = count;
                    res.clear();
                    res.push_back(root->val);
                }

                root = root->right;
            }
        }
        return res;
    }
};
// @lc code=end

